lib/chef/node_map.rb

```diff
@@ -59,6 +65,31 @@ def set(key, klass, platform: nil, platform_version: nil, platform_family: nil,
       new_matcher[:block] = block if block
       new_matcher[:canonical] = canonical if canonical
       new_matcher[:override] = override if override
+      new_matcher[:cookbook_override] = allow_cookbook_override
+      new_matcher[:core_override] = __core_override__
+
+      # Check if the key is already present and locked, unless the override is allowed.
+      # The checks to see if we should reject, in order:
+      # 1. Core overide mode is not set.
+      # 2. The key exists.
+      # 3. At least one previous `provides` is now locked.
+      # 4. No previous `provides` had `allow_cookbook_override`, either set to
```

Comment:
I'm not sure I follow this logic.

Lets be concrete:

* Package::YumPackage (builtin) is locked, and has `allow_cookbook_override
* Package::HonkyDory (builtin) is locked, and does NOT have `allow_cookbook_override`
* We now hit Package::WillyNilly - should it be able  override HonkDory even though HonkyDory has not allowed overrides?

Since only internal stuff can be `locked` (I think?), shouldn't we ensure that all things which are locked also allow overrides? Or am I misunderstanding?

Reply:
The locking happens all at once after all of core's resource and provider files are loaded, so anything in core can always do whatever it wants. Only after https://github.com/chef/chef/pull/7224/files#diff-c34d08382bb25f61e44ecbd5f50f4166R278 do we transition to the new rules. So if `Package::WillyNilly` was in a cookbook, then it would be rejected starting in 15.0 unless it uses `__core_override__`.

Reply:
Oh I think I see what you mean, the rules around `allow_cookbook_override` make less sense for providers (which I think is what you mean here?), setting that on just one of the core package providers would indeed get funky.